File: Python/GoBackN.py
import time

class Frame:
    def init(self, seq_num, data):
        self.seq_num = seq_num
        self.data = data

def send_frame(frame):
    print(f"Sending frame {frame.seq_num}: {frame.data}")
    time.sleep(1)  # Simulate transmission delay
    return frame

def receive_ack(seq_num):
    ack = input(f"Did you receive ACK for frame {seq_num}? (yes/no): ").lower()
    if ack == 'yes':
        print(f"ACK for frame {seq_num} received.")
        return True
    else:
        print(f"ACK for frame {seq_num} not received.")
        return False

def go_back_n(frames, window_size):
    n = len(frames)
    base = 0
    next_seq_num = 0
    while base < n:
        while next_seq_num < base + window_size and next_seq_num < n:
            frame = Frame(next_seq_num, frames[next_seq_num])
            send_frame(frame)
            next_seq_num += 1

        ack_received = True
        for i in range(base, next_seq_num):
            if not receive_ack(i):
                ack_received = False
                break

        if ack_received:
            base = next_seq_num
        else:
            print(f"Resending window starting from frame {base}")
            next_seq_num = base

def main():
    frames = []
    while True:
        frame_data = input("Enter frame data (or 'done' to finish): ")
        if frame_data.lower() == 'done':
            break
        frames.append(frame_data)

    window_size = int(input("Enter the window size: "))
    go_back_n(frames, window_size)

# Directly call the main function
main()

File: C/GoBackN.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

typedef struct {
    int seq_num;
    char data[100];
} Frame;

void send_frame(Frame frame) {
    printf("Sending frame %d: %s\n", frame.seq_num, frame.data);
    sleep(1);  // Simulate transmission delay
}

int receive_ack(int seq_num) {
    char response[10];
    printf("Did you receive ACK for frame %d? (yes/no): ", seq_num);
    scanf("%s", response);
    if (strcmp(response, "yes") == 0) {
        printf("ACK for frame %d received.\n", seq_num);
        return 1;
    } else {
        printf("ACK for frame %d not received.\n", seq_num);
        return 0;
    }
}

void go_back_n(Frame *frames, int frame_count, int window_size) {
    int base = 0;
    int next_seq_num = 0;

    while (base < frame_count) {
        while (next_seq_num < base + window_size && next_seq_num < frame_count) {
            send_frame(frames[next_seq_num]);
            next_seq_num++;
        }

        int ack_received = 1;
        for (int i = base; i < next_seq_num; i++) {
            if (!receive_ack(i)) {
                ack_received = 0;
                next_seq_num = i + 1;  // Start sending from the lost frame
                break;
            }
        }

        if (ack_received) {
            base = next_seq_num;
        } else {
            printf("Resending window starting from frame %d\n", base);
        }
    }
}

int main() {
    int frame_count, window_size;

    printf("Enter the number of frames: ");
    scanf("%d", &frame_count);

    Frame *frames = (Frame *)malloc(frame_count * sizeof(Frame));

    for (int i = 0; i < frame_count; i++) {
        printf("Enter data for frame %d: ", i);
        scanf("%s", frames[i].data);
        frames[i].seq_num = i;
    }

    printf("Enter the window size: ");
    scanf("%d", &window_size);

    go_back_n(frames, frame_count, window_size);

    free(frames);
    return 0;
}